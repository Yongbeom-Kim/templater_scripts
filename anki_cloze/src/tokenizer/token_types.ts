export enum TokenType {
  Text = "text",
  Number = "number",  
  Punctuation = "punctuation",
  Newline = "newline", // one or more consecutive newlines
  Whitespace = "whitespace" // one or more consecutive spaces or tabs
}

export type Token = {
  type: TokenType;
  lexeme: string;
  line_pos: number;
  col_pos: number;
}

export namespace Token {
  export const Create = (state: TokenizerState, type: TokenType, lexeme: string): Token => {
    return {
      type,
      lexeme,
      line_pos: state.line_pos,
      col_pos: state.col_pos
    }
  }
}
export class TokenizerState {
  constructor(
    readonly text: string,
    readonly next: number = 0,
    readonly line_pos: number = 1,
    readonly col_pos: number = 0
  ) {
  }
  
  eof(): boolean {
    return this.next >= this.text.length;
  }

  good(): boolean {
    return this.next < this.text.length;
  }

  peek(n_chars: number): string {
    return this.consume(n_chars)[1];
  }

  private getUpdatedPosition(textConsumed: string): [number, number] {
    let newLinePos = this.line_pos;
    let newColPos = this.col_pos;
    if (!textConsumed.includes("\n")) {
      newColPos += textConsumed.length;
    } else {
      const newLines = textConsumed.split("\n").length - 1;
      newLinePos += newLines;
      newColPos = textConsumed.length - textConsumed.lastIndexOf("\n") - 1;
    }
    return [newLinePos, newColPos];
  }

  consume(n_chars: number): [TokenizerState, string] {
    const newText = this.text.slice(this.next, this.next + n_chars);
    const [newLinePos, newColPos] = this.getUpdatedPosition(newText);
    return [new TokenizerState(this.text, this.next + n_chars, newLinePos, newColPos), newText];
  }

  peekUntil(predicate: (char: string) => boolean, max_chars: number = -1): string {
    return this.consumeUntil(predicate, max_chars)[1];
  }

  consumeUntil(predicate: (char: string) => boolean, max_chars: number = -1): [TokenizerState, string] {
    let result = "";
    let idx = this.next;
    while (max_chars != 0 && idx < this.text.length && !predicate(this.text[idx])) {
      result += this.text[idx];
      idx++;
      max_chars--;
    }
    const [newLinePos, newColPos] = this.getUpdatedPosition(result);
    return [new TokenizerState(this.text, idx, newLinePos, newColPos), result];
  }

  isStartOfLine(): boolean {
    return this.next == 0 || this.text[this.next - 1] == "\n";
  }
}