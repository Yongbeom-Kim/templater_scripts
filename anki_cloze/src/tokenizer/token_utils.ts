export const isSpace = (char: string): boolean => {
  return char == " ";
};

export const isNewline = (char: string): boolean => {
  return char == "\n";
};
