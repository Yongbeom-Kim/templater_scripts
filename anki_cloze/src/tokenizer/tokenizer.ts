import { Token, TokenizerState, TokenType } from "./token_types";

// TODO: Consider converting to stream-based tokenizer if performance becomes an issue
export const tokenize = (text: string): Token[] => {
  // Normalize newlines
  text = text.replaceAll("\r\n", "\n").replaceAll("\r", "\n");

  const tokens: Token[] = [];
  let state = new TokenizerState(text);
  while (state.good()) {
    let lexeme;
    let updatedState;
    if (isText(state.peek(1))) {
      [updatedState, lexeme] = state.consumeUntil(not(isText));
      tokens.push(Token.Create(state, TokenType.Text, lexeme));
    } else if (isDigit(state.peek(1))) {
      [updatedState, lexeme] = state.consumeUntil(not(isDigit));
      tokens.push(Token.Create(state, TokenType.Number, lexeme));
    } else if (isSpace(state.peek(1))) {
      [updatedState, lexeme] = state.consumeUntil(not(isSpace));
      tokens.push(Token.Create(state, TokenType.Whitespace, lexeme));
    } else if (isNewline(state.peek(1))) {
      [updatedState, lexeme] = state.consume(1);
      tokens.push(Token.Create(state, TokenType.Newline, lexeme));
    } else if (isPunctuation(state.peek(1))) {
      [updatedState, lexeme] = state.consumeUntil(not(isPunctuation));
      tokens.push(Token.Create(state, TokenType.Punctuation, lexeme));
    } else {
      throw new Error(`Unexpected character: "${state.peek(1)}". Index: (${state.line_pos}, ${state.col_pos}). Context: ${state.peek(5)}...`);
    }
    state = updatedState;
  }
  return tokens;
};

const isText = (c: string): boolean => {
  return !!c.match(/[a-zA-Z_]/);
};

const isDigit = (c: string): boolean => {
  return !!c.match(/\d/);
};

const isSpace = (c: string): boolean => {
  // \u00A0 is a non-breaking space
  return c == " " || c == "\t" || c == "\u00A0";
};

const isNewline = (c: string): boolean => {
  return c == "\n";
};

const isPunctuation = (c: string): boolean => {
  return !!c.match(/[^\w\s]/);
};

const not = (predicate: (...args: any[]) => boolean) => {
  return (...args: any[]) => !predicate(...args);
};
