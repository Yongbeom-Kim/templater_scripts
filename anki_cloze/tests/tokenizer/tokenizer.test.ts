import { describe, test, expect } from "@jest/globals";
import { tokenize } from "../../src/tokenizer/tokenizer";
import { TokenType } from "../../src/tokenizer/token_types";
import { TestCases as ListTestCases } from "../data/list.testcase";
import { TestCases as TextTestCases } from "../data/text.testcase";

describe("test text.testcase.ts", () => {
  it.each(TextTestCases)(
    'should tokenize "%p" correctly',
    ({ input, tokenize_expect }) => {
      const tokens = tokenize(input);
      tokenize_expect(tokens);
      expect(tokens).toMatchSnapshot();
    },
  );
});

describe("test list.testcase.ts", () => {
  it.each(ListTestCases)(
    'should tokenize "%p" correctly',
    ({ input, tokenize_expect }) => {
      const tokens = tokenize(input);
      tokenize_expect(tokens);
      expect(tokens).toMatchSnapshot();
    },
  );
});
