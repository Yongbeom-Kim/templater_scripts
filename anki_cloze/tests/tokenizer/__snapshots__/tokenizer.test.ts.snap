// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test list.testcase.ts should tokenize "{"clozify_expect": [Function clozify_expect], "description": "Parse nested list of items", "input": "- Front - Back
  - Front = Back
1. Front - Back
  1. Front = Back
  2. Front = Back", "main_expect": [Function main_expect], "parse_expect": [Function parse_expect], "tokenize_expect": [Function tokenize_expect]}" correctly 1`] = `
[
  {
    "col_pos": 0,
    "lexeme": "-",
    "line_pos": 1,
    "type": "punctuation",
  },
  {
    "col_pos": 1,
    "lexeme": " ",
    "line_pos": 1,
    "type": "whitespace",
  },
  {
    "col_pos": 2,
    "lexeme": "Front",
    "line_pos": 1,
    "type": "text",
  },
  {
    "col_pos": 7,
    "lexeme": " ",
    "line_pos": 1,
    "type": "whitespace",
  },
  {
    "col_pos": 8,
    "lexeme": "-",
    "line_pos": 1,
    "type": "punctuation",
  },
  {
    "col_pos": 9,
    "lexeme": " ",
    "line_pos": 1,
    "type": "whitespace",
  },
  {
    "col_pos": 10,
    "lexeme": "Back",
    "line_pos": 1,
    "type": "text",
  },
  {
    "col_pos": 14,
    "lexeme": "
",
    "line_pos": 1,
    "type": "newline",
  },
  {
    "col_pos": 0,
    "lexeme": "  ",
    "line_pos": 2,
    "type": "whitespace",
  },
  {
    "col_pos": 2,
    "lexeme": "-",
    "line_pos": 2,
    "type": "punctuation",
  },
  {
    "col_pos": 3,
    "lexeme": " ",
    "line_pos": 2,
    "type": "whitespace",
  },
  {
    "col_pos": 4,
    "lexeme": "Front",
    "line_pos": 2,
    "type": "text",
  },
  {
    "col_pos": 9,
    "lexeme": " ",
    "line_pos": 2,
    "type": "whitespace",
  },
  {
    "col_pos": 10,
    "lexeme": "=",
    "line_pos": 2,
    "type": "punctuation",
  },
  {
    "col_pos": 11,
    "lexeme": " ",
    "line_pos": 2,
    "type": "whitespace",
  },
  {
    "col_pos": 12,
    "lexeme": "Back",
    "line_pos": 2,
    "type": "text",
  },
  {
    "col_pos": 16,
    "lexeme": "
",
    "line_pos": 2,
    "type": "newline",
  },
  {
    "col_pos": 0,
    "lexeme": "1",
    "line_pos": 3,
    "type": "number",
  },
  {
    "col_pos": 1,
    "lexeme": ".",
    "line_pos": 3,
    "type": "punctuation",
  },
  {
    "col_pos": 2,
    "lexeme": " ",
    "line_pos": 3,
    "type": "whitespace",
  },
  {
    "col_pos": 3,
    "lexeme": "Front",
    "line_pos": 3,
    "type": "text",
  },
  {
    "col_pos": 8,
    "lexeme": " ",
    "line_pos": 3,
    "type": "whitespace",
  },
  {
    "col_pos": 9,
    "lexeme": "-",
    "line_pos": 3,
    "type": "punctuation",
  },
  {
    "col_pos": 10,
    "lexeme": " ",
    "line_pos": 3,
    "type": "whitespace",
  },
  {
    "col_pos": 11,
    "lexeme": "Back",
    "line_pos": 3,
    "type": "text",
  },
  {
    "col_pos": 15,
    "lexeme": "
",
    "line_pos": 3,
    "type": "newline",
  },
  {
    "col_pos": 0,
    "lexeme": "  ",
    "line_pos": 4,
    "type": "whitespace",
  },
  {
    "col_pos": 2,
    "lexeme": "1",
    "line_pos": 4,
    "type": "number",
  },
  {
    "col_pos": 3,
    "lexeme": ".",
    "line_pos": 4,
    "type": "punctuation",
  },
  {
    "col_pos": 4,
    "lexeme": " ",
    "line_pos": 4,
    "type": "whitespace",
  },
  {
    "col_pos": 5,
    "lexeme": "Front",
    "line_pos": 4,
    "type": "text",
  },
  {
    "col_pos": 10,
    "lexeme": " ",
    "line_pos": 4,
    "type": "whitespace",
  },
  {
    "col_pos": 11,
    "lexeme": "=",
    "line_pos": 4,
    "type": "punctuation",
  },
  {
    "col_pos": 12,
    "lexeme": " ",
    "line_pos": 4,
    "type": "whitespace",
  },
  {
    "col_pos": 13,
    "lexeme": "Back",
    "line_pos": 4,
    "type": "text",
  },
  {
    "col_pos": 17,
    "lexeme": "
",
    "line_pos": 4,
    "type": "newline",
  },
  {
    "col_pos": 0,
    "lexeme": "  ",
    "line_pos": 5,
    "type": "whitespace",
  },
  {
    "col_pos": 2,
    "lexeme": "2",
    "line_pos": 5,
    "type": "number",
  },
  {
    "col_pos": 3,
    "lexeme": ".",
    "line_pos": 5,
    "type": "punctuation",
  },
  {
    "col_pos": 4,
    "lexeme": " ",
    "line_pos": 5,
    "type": "whitespace",
  },
  {
    "col_pos": 5,
    "lexeme": "Front",
    "line_pos": 5,
    "type": "text",
  },
  {
    "col_pos": 10,
    "lexeme": " ",
    "line_pos": 5,
    "type": "whitespace",
  },
  {
    "col_pos": 11,
    "lexeme": "=",
    "line_pos": 5,
    "type": "punctuation",
  },
  {
    "col_pos": 12,
    "lexeme": " ",
    "line_pos": 5,
    "type": "whitespace",
  },
  {
    "col_pos": 13,
    "lexeme": "Back",
    "line_pos": 5,
    "type": "text",
  },
]
`;

exports[`test text.testcase.ts should tokenize "{"clozify_expect": [Function clozify_expect], "description": "Mixed whitespace and newlines", "input": " 	 
 	 ", "main_expect": [Function main_expect], "parse_expect": [Function parse_expect], "tokenize_expect": [Function tokenize_expect]}" correctly 1`] = `
[
  {
    "col_pos": 0,
    "lexeme": " 	 ",
    "line_pos": 1,
    "type": "whitespace",
  },
  {
    "col_pos": 3,
    "lexeme": "
",
    "line_pos": 1,
    "type": "newline",
  },
  {
    "col_pos": 0,
    "lexeme": " 	 ",
    "line_pos": 2,
    "type": "whitespace",
  },
]
`;

exports[`test text.testcase.ts should tokenize "{"clozify_expect": [Function clozify_expect], "description": "Multiple newlines", "input": "

", "main_expect": [Function main_expect], "parse_expect": [Function parse_expect], "tokenize_expect": [Function tokenize_expect]}" correctly 1`] = `
[
  {
    "col_pos": 0,
    "lexeme": "

",
    "line_pos": 1,
    "type": "newline",
  },
]
`;

exports[`test text.testcase.ts should tokenize "{"clozify_expect": [Function clozify_expect], "description": "Simple text with newline", "input": "Hello 123
World", "main_expect": [Function main_expect], "parse_expect": [Function parse_expect], "tokenize_expect": [Function tokenize_expect]}" correctly 1`] = `
[
  {
    "col_pos": 0,
    "lexeme": "Hello",
    "line_pos": 1,
    "type": "text",
  },
  {
    "col_pos": 5,
    "lexeme": " ",
    "line_pos": 1,
    "type": "whitespace",
  },
  {
    "col_pos": 6,
    "lexeme": "123",
    "line_pos": 1,
    "type": "number",
  },
  {
    "col_pos": 9,
    "lexeme": "
",
    "line_pos": 1,
    "type": "newline",
  },
  {
    "col_pos": 0,
    "lexeme": "World",
    "line_pos": 2,
    "type": "text",
  },
]
`;

exports[`test text.testcase.ts should tokenize "{"clozify_expect": [Function clozify_expect], "description": "Simple text with numbers", "input": "123 456", "main_expect": [Function main_expect], "parse_expect": [Function parse_expect], "tokenize_expect": [Function tokenize_expect]}" correctly 1`] = `
[
  {
    "col_pos": 0,
    "lexeme": "123",
    "line_pos": 1,
    "type": "number",
  },
  {
    "col_pos": 3,
    "lexeme": " ",
    "line_pos": 1,
    "type": "whitespace",
  },
  {
    "col_pos": 4,
    "lexeme": "456",
    "line_pos": 1,
    "type": "number",
  },
]
`;

exports[`test text.testcase.ts should tokenize "{"clozify_expect": [Function clozify_expect], "description": "Text with mixed whitespace and newlines", "input": "Hello 123
 	World
456", "main_expect": [Function main_expect], "parse_expect": [Function parse_expect], "tokenize_expect": [Function tokenize_expect]}" correctly 1`] = `
[
  {
    "col_pos": 0,
    "lexeme": "Hello",
    "line_pos": 1,
    "type": "text",
  },
  {
    "col_pos": 5,
    "lexeme": " ",
    "line_pos": 1,
    "type": "whitespace",
  },
  {
    "col_pos": 6,
    "lexeme": "123",
    "line_pos": 1,
    "type": "number",
  },
  {
    "col_pos": 9,
    "lexeme": "
",
    "line_pos": 1,
    "type": "newline",
  },
  {
    "col_pos": 0,
    "lexeme": " 	",
    "line_pos": 2,
    "type": "whitespace",
  },
  {
    "col_pos": 2,
    "lexeme": "World",
    "line_pos": 2,
    "type": "text",
  },
  {
    "col_pos": 7,
    "lexeme": "
",
    "line_pos": 2,
    "type": "newline",
  },
  {
    "col_pos": 0,
    "lexeme": "456",
    "line_pos": 3,
    "type": "number",
  },
]
`;

exports[`test text.testcase.ts should tokenize "{"clozify_expect": [Function clozify_expect], "description": "Text with multiple newlines", "input": "Multiple


Newlines


", "main_expect": [Function main_expect], "parse_expect": [Function parse_expect], "tokenize_expect": [Function tokenize_expect]}" correctly 1`] = `
[
  {
    "col_pos": 0,
    "lexeme": "Multiple",
    "line_pos": 1,
    "type": "text",
  },
  {
    "col_pos": 8,
    "lexeme": "


",
    "line_pos": 1,
    "type": "newline",
  },
  {
    "col_pos": 0,
    "lexeme": "Newlines",
    "line_pos": 4,
    "type": "text",
  },
  {
    "col_pos": 8,
    "lexeme": "


",
    "line_pos": 4,
    "type": "newline",
  },
]
`;

exports[`test text.testcase.ts should tokenize "{"clozify_expect": [Function clozify_expect], "description": "Whitespace characters", "input": " 	 ", "main_expect": [Function main_expect], "parse_expect": [Function parse_expect], "tokenize_expect": [Function tokenize_expect]}" correctly 1`] = `
[
  {
    "col_pos": 0,
    "lexeme": " 	 ",
    "line_pos": 1,
    "type": "whitespace",
  },
]
`;
